### PR TITLE
Support AD Password Complexity

### PR DESCRIPTION
**Pull request by trevoring-okta**
## Description:
- SIW-side changes for supporting AD password complexity, which is a new password requirement that forces users to have 3 out of 4 of the following: lowercase letter, uppercase letter, number, symbol
- Enforces the AD complexity requirements in gen 3's client-side validation logic (gen 2's password validation is only server-side)
- Adds new mocks that contain the `useADComplexityRequirements` option for all e2e flows that display password requirements

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-637006](https://oktainc.atlassian.net/browse/OKTA-637006)
- [Corresponding backend/admin UI ticket](https://oktainc.atlassian.net/browse/OKTA-625930)
- [Corresponding backend/admin UI PR](https://github.com/atko-eng/okta-core/pull/83514)
- [Bacon Link](https://bacon-go.aue1e.saasure.net/commits?artifact=okta-signin-widget&branch=ti-OKTA-637006-ad-pw-complexity&page=1&pageSize=6&sha=155ceef0377ef71578e5341a079bccc2afd154b2&tab=main)

#### Gen 2 AD Complexity Displayed as Password Requirement
https://github.com/okta/okta-signin-widget/assets/106110543/dbfbf8e2-b2eb-4d31-ad52-34a0ba214de1

#### Gen 3 AD Complexity Displayed as Password Requirement and Validated
https://github.com/okta/okta-signin-widget/assets/106110543/c0771bc0-5421-48eb-926c-7034121f8827

### Downstream Monolith Build:
- [Downstream Link](https://bacon-go.aue1e.saasure.net/commits?artifact=monolith&page=1&pageSize=6&sha=656aaf133be21c3b73f3492301f5d3683f53f0d7&tab=main)

### CODE REVIEW

File: src/util/FactorUtil.js [hunk 1]
```diff
@@ -178,9 +178,19 @@ const getPasswordComplexityRequirementsAsArray = function(policy, i18nKeys) {
...
+    // If useADComplexityRequirements is true, ignore casing, number, and symbol rules since
+    // AD validator handles those requirements
+    if (policyComplexity.useADComplexityRequirements) {
+      const allowed = ['minLength', 'useADComplexityRequirements', 'excludeUsername', 'excludeFirstName',
+        'excludeLastName', 'excludeAttributes'];
+      filteredPolicyComplexity = _.pick(policyComplexity, allowed);
+    }
 
-    const requirements = _.map(policyComplexity, function(complexityValue, complexityType) {
+    const requirements = _.map(filteredPolicyComplexity, function(complexityValue, complexityType) {
```
Comment (trevoring-okta) [inline src/util/FactorUtil.js:193]: Gen 2 changes only require filtering out the complexity fields to render if `useADComplexityRequirements` is present

File: src/v3/src/util/passwordUtils.ts [hunk 2]
```diff
@@ -136,6 +137,24 @@ const excludeLastNameValidator = (
...
+  const requiredValidators: ((password: string, limit: unknown) => boolean)[] = [
+    minLowerCaseValidator,
+    minUpperCaseValidator,
+    minNumberValidator,
+    minSymbolValidator,
+  ];
+
+  // Runs the 4 validation functions and returns the number that passed
+  const numRequirementsMet = requiredValidators.reduce((num, validator) => (
+    validator(password, 1) ? num + 1 : num), 0);
```
Comment (trevoring-okta) [inline src/v3/src/util/passwordUtils.ts:152]: Calls each of the 4 possible validators that can satisfy AD requirements with a hard-coded `limit=1`

**Comment by jakebacher2-okta** [comment]
Code changes LGTM but a couple questions:
can you post a video of the V2 changes as well? Including if there's an error 
And for the v3 changes, is the error message supposed to disappear once you fulfill the requirements? - I haven't worked wit V3 very much, so I'm not sure if this is expected behavior.

**Comment by trevoring-okta** [comment]
@jakebacher2-okta 
- Updated with video of v2 changes with error. Since we don't do client-side validation all you should see is this generic "We found some errors..." global error and the field level "Password requirements were not met" when combining these changes with your backend work
- Regarding the error message disappearing, yeah the message disappears `onBlur` it just didn't disappear in the video because I didn't remove focus from the password field

**Review by jakebacher2-okta — APPROVED** [review]
LGTM fro auth-experiences side! I'd recommend getting an approval from your team as well, just since I don't work with the SIW too often and there might be something I missed

File: src/v3/src/transformer/password/passwordSettingsUtils.test.ts [hunk 3]
```diff
@@ -81,4 +81,27 @@ describe('getComplexityItems', () => {
...
+      minLength: 8,
+      minLowerCase: 1,
+      minUpperCase: 1,
+      minNumber: 1,
+      minSymbol: 1,
+      excludeUsername: true,
+      excludeFirstName: true,
+      excludeLastName: true,
+      useADComplexityRequirements: true,
+    } as unknown as ComplexityRequirements;
```
[the remark this answers is not in the document]
Comment (trevoring-okta) [reply src/v3/src/transformer/password/passwordSettingsUtils.test.ts:96]: Just checked and the cast isn't necessary, it was just copied from the declaration of `complexity` for the test above

File: src/v3/src/transformer/password/passwordSettingsUtils.ts [hunk 4]
```diff
@@ -39,7 +39,16 @@ export const getComplexityItems = (complexity?: ComplexityRequirements): ListIte
...
   }
 
-  Object.entries(complexity).forEach(([key, value]) => {
+  let filteredComplexity: ComplexityRequirements = complexity;
+
+  // If useADComplexityRequirements is true, ignore casing, number, and symbol rules since AD validator handles those requirements
+  if (complexity.useADComplexityRequirements) {
+    const allowed = ['minLength', 'useADComplexityRequirements', 'excludeUsername', 'excludeFirstName',
+      'excludeLastName', 'excludeAttributes'];
+    filteredComplexity = Object.fromEntries(allowed.map((key) => [key, complexity[key]]));
```
[the remark this answers is not in the document]
Comment (trevoring-okta) [reply src/v3/src/transformer/password/passwordSettingsUtils.ts:48]: Not sure, will switch to building object to be safe

File: playground/mocks/config/responseConfig.js [hunk 5]
```diff
@@ -33,7 +33,7 @@ const idx = {
     // 'error-identify-with-only-one-third-party-idp',
     // 'authenticator-enroll-email',
     // 'error-internal-server-error',
-    // 'authenticator-enroll-password',
+    'authenticator-enroll-password-with-ad-req',
```
Comment (lesterchoi-okta) [inline playground/mocks/config/responseConfig.js:36]: Mind un-commenting "identify" above and commenting this one out? We usually don't commit changes that affect the mock response cycle.

```suggestion
    // 'authenticator-enroll-password-with-ad-req',
```

File: src/v3/src/util/passwordUtils.test.ts [hunk 6]
```diff
@@ -278,6 +278,73 @@ describe('PasswordUtils Tests', () => {
       .toEqual(true);
   });
 
+  describe('Validate useADComplexityRequirements', () => {
+    it('should validate password for at least 3 of the following: lowercase letter, uppercase letter, number, symbol', () => {
+      expect(validatePassword('abcd1234', userInfo, { complexity: { useADComplexityRequirements: true } })
+        ?.useADComplexityRequirements)
+        .toEqual(false);
+      expect(validatePassword('aaaaaaaa', userInfo, { complexity: { useADComplexityRequirements: true } })
```
Comment (lesterchoi-okta) [inline src/v3/src/util/passwordUtils.test.ts:286]: nit: consider splitting into separate "it" cases so it's clear when one fails

File: src/v3/src/util/passwordUtils.test.ts [hunk 7]
```diff
@@ -278,6 +278,73 @@ describe('PasswordUtils Tests', () => {
       .toEqual(true);
   });
 
+  describe('Validate useADComplexityRequirements', () => {
+    it('should validate password for at least 3 of the following: lowercase letter, uppercase letter, number, symbol', () => {
```
Comment (lesterchoi-okta) [inline src/v3/src/util/passwordUtils.test.ts:282]: https://jestjs.io/docs/api#testeachtablename-fn-timeout

**Review by lesterchoi-okta — APPROVED** [review]

**Review by lesterchoi-okta — APPROVED** [review]

File: playground/mocks/config/responseConfig.js (hunk 5 above)
Comment (trevoring-okta) [reply playground/mocks/config/responseConfig.js:36]: Oh yeah this was accidental, thanks for catching it

**Pull request merged**